### PR TITLE
Implement AxivionSuite tool which queries the Axivion Dashboard for violations

### PR DESCRIPTION
**Pull request by arturbosch**

### CODE REVIEW

File: src/main/java/io/jenkins/plugins/analysis/warnings/axivion/AxivionSuite.java [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+import java.nio.file.InvalidPathException;
+import java.nio.file.Paths;
+import java.util.Collections;
+import java.util.List;
+import javax.annotation.Nonnull;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.http.auth.UsernamePasswordCredentials;
+
+import com.cloudbees.plugins.credentials.CredentialsMatchers;
```
Comment (uhafner) [inline src/main/java/io/jenkins/plugins/analysis/warnings/axivion/AxivionSuite.java]: Where is this dependency coming from? Jenkins-core?

File: src/main/java/io/jenkins/plugins/analysis/warnings/axivion/AxivionSuite.java [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+
+import io.jenkins.plugins.analysis.core.model.Tool;
+import io.jenkins.plugins.analysis.core.util.EnvironmentResolver;
+import io.jenkins.plugins.analysis.core.util.LogHandler;
+
+/** Provides a parser and customized messages for the Axivion Suite. */
+public class AxivionSuite extends Tool {
+
+    private static final long serialVersionUID = 8670135727302169818L;
+    static final String ID = "AxivionStoppingSoftwareErosion";
```
Comment (uhafner) [inline src/main/java/io/jenkins/plugins/analysis/warnings/axivion/AxivionSuite.java]: quite a long URL, is this intended?

File: src/test/java/io/jenkins/plugins/analysis/warnings/axivion/AxivionParserTest.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        String[] expected = {
+                "MisraC++-7.1.1",
+                "Generic-NamingConvention",
+                "Style.Indentation-AllmanBraces"
+        };
+        for (final Issue actual : report) {
+            assertTrue(ArrayUtils.contains(expected, actual.getType()));
+        }
+    }
+}
```
Comment (uhafner) [inline src/test/java/io/jenkins/plugins/analysis/warnings/axivion/AxivionParserTest.java]: Some more tests would be helpful.

File: src/test/java/io/jenkins/plugins/analysis/warnings/axivion/AxivionParserTest.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        );
+
+        String[] expected = {
+                "MisraC++-7.1.1",
+                "Generic-NamingConvention",
+                "Style.Indentation-AllmanBraces"
+        };
+        for (final Issue actual : report) {
+            assertTrue(ArrayUtils.contains(expected, actual.getType()));
+        }
```
Comment (uhafner) [inline src/test/java/io/jenkins/plugins/analysis/warnings/axivion/AxivionParserTest.java]: check all properties of an issue

File: src/test/java/io/jenkins/plugins/analysis/warnings/axivion/AxivionParserTest.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+class AxivionParserTest {
+
+    @Test
+    void parseAxivionIssuesViaDashboardJsonReport() throws IOException {
+        AxivionParser parser = new AxivionParser("testUrl", null, "testDir");
+        URL svJson = this.getClass().getResource("/io/jenkins/plugins/analysis/warnings/axivion/sv.json");
+        final String payload = new String(new Resource(svJson).asByteArray());
+
+        Report report = new Report();
+        parser.processIssues(
```
Comment (uhafner) [inline src/test/java/io/jenkins/plugins/analysis/warnings/axivion/AxivionParserTest.java]: why don't you test the `parse()` method?

File: src/main/java/io/jenkins/plugins/analysis/warnings/axivion/AxivionSuite.java (hunk 2 above)
Comment (arturbosch) [reply src/main/java/io/jenkins/plugins/analysis/warnings/axivion/AxivionSuite.java]: Yep, it was intented, but AxivionSuite is ok too. Changed it.

File: src/main/java/io/jenkins/plugins/analysis/warnings/axivion/AxivionSuite.java (hunk 1 above)
Comment (arturbosch) [reply src/main/java/io/jenkins/plugins/analysis/warnings/axivion/AxivionSuite.java]: It comes from the `Credentials Plugin` which is shipped by Jenkins and can't be deinstalled.

**Comment by arturbosch** [comment]
Thanks for the review, I incorporated your comments.

What is the right way to integrate jelly templating?
When this plugin was a standalone plugin jelly worked, I had a hard time migrating the jelly files to the new resources package structure. Lastly it worked, now after renaming the ID field, my custom three input fields again disappeared.
The parser still works for already configured projects.

I can't reproduce the architecture violations locally.
On CI it tells me not to use `Jenkins.getInstance()`. What is the prefered way to retrieve the credentials ids from jenkins?

Is `assertAll` not allowed in your code base due to longer CI runs?

File: src/main/java/io/jenkins/plugins/analysis/warnings/axivion/AxivionSuite.java (hunk 2 above)
Comment (uhafner) [reply src/main/java/io/jenkins/plugins/analysis/warnings/axivion/AxivionSuite.java]: It is used as URL to the results, so typically it does contain only lowercase characters (separated by -), see https://github.com/jenkinsci/warnings-ng-plugin/blob/master/SUPPORTED-FORMATS.md.

**Comment by uhafner** [comment]
{quote}
What is the right way to integrate jelly templating?
When this plugin was a standalone plugin jelly worked, I had a hard time migrating the jelly files to the new resources package structure. Lastly it worked, now after renaming the ID field, my custom three input fields again disappeared.
The parser still works for already configured projects.
{quote}

I need to check that locally... 

{quote}
I can't reproduce the architecture violations locally.
On CI it tells me not to use Jenkins.getInstance(). What is the prefered way to retrieve the credentials ids from jenkins?
{quote}

The architecture violations are from a simple unit test, why does it not work locally?

I'm currently preparing the project to be used in my testing course here at the university, so I want that everybody conforms to my guidelines. I hope they are not too academic;-) Two new rules that are now enforced with an architecture test:

I generally want to forbid direct `Jenkins.getInstance` from code since these code parts cannot be unit tested (you need an integration test). In order to make auch a code testable you need to call the Jenkins façade that handles these calls. The Jenkins façade can be replaced in a unit test with a stub. If your method to Jenkins instance is not available yet, please add it to the façade. Here is an example: https://github.com/jenkinsci/warnings-ng-plugin/blob/master/src/test/java/io/jenkins/plugins/analysis/core/util/ModelValidationTest.java 

{quote}
Is assertAll not allowed in your code base due to longer CI runs?
{quote}

No, I simply want to use only one assertion framework in my modules and that is *AssertJ*. Otherwise contributors get confused if some use JUnit4, some use JUnit5, some AssertJ, etc.

File: src/main/resources/io/jenkins/plugins/analysis/warnings/axivion/local-config.jelly [hunk 6]
```diff
@@ -0,0 +1,18 @@
+<?jelly escape-by-default='true'?>
```
Comment (uhafner) [inline src/main/resources/io/jenkins/plugins/analysis/warnings/axivion/local-config.jelly]: This file needs to be renamed to config.jelly! local-config.jelly is only for subtypes of `ReportParsingTool`.

**Comment by arturbosch** [comment]
Thanks for the input, `resources/AxivionSuite/config.jelly` and `ID=axivion-suite` made it work.

**Comment by uhafner** [comment]
Thanks!

**Pull request merged**